### PR TITLE
Signal processing (FIR/IIR filtering) content

### PR DESCRIPTION
**Pull request by WarrenWeckesser**

### CODE REVIEW

File: papers/scipy/code/signal/bandpass_batch_example.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
+from scipy.signal import sosfreqz, butter, sosfilt
+import matplotlib.pyplot as plt
+from butter import butter_bandpass, butter_bandpass_filt
```
Comment (rgommers) [inline papers/scipy/code/signal/bandpass_batch_example.py]: ``butter_bandpass_filt`` unused

File: papers/scipy/code/signal/bandpass_batch_example.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
+from scipy.signal import sosfreqz, butter, sosfilt
```
Comment (rgommers) [inline papers/scipy/code/signal/bandpass_batch_example.py]: ``sosfreqz, butter,`` unused

File: papers/scipy/code/signal/bandpass_batch_example.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+plt.figure(figsize=(4.0, 3.2))
+
+plt.plot(t, x, 'k', alpha=0.4, linewidth=1, label='Noisy signal')
+
+start = 0
+alpha = 0.5
+while start < len(x):
+    stop = min(start + batch_size, len(x))
+    if start == 0 or start == batch_size:
+        lbl = 'Filtered signal'
```
Comment (rgommers) [inline papers/scipy/code/signal/bandpass_batch_example.py]: nitpick: ``lbl`` is better named ``label``

File: papers/scipy/code/signal/bandpass_example.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import numpy as np
+from scipy.signal import sosfreqz, butter, sosfilt
```
Comment (rgommers) [inline papers/scipy/code/signal/bandpass_example.py]: ``butter, sosfilt`` unused

File: papers/scipy/code/signal/bandpass_batch_example.py (hunk 3 above)
Comment (rgommers) [reply papers/scipy/code/signal/bandpass_batch_example.py]: more importantly, the label ``Filtered signal`` now appears twice in the plot (light and dark blue), it would be better to only let it appear for dark blue (looks a bit odd to me now). The figure caption explains the rest.

File: papers/scipy/code/signal/fir_filter_design_examples.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+taps_none = firwin2(numtaps, freqs, gains, nyq=0.5*fs, window=None)
+taps_fw = firwin2(numtaps, freqs, gains, nyq=0.5*fs)
+
+atten = kaiser_atten(numtaps, 20/(0.5*fs))
+beta = kaiser_beta(atten)
+taps_k = firwin2(numtaps, freqs, gains, nyq=0.5*fs, window=('kaiser', beta))
+
+#bands = [0, 5, 5, 25, 25, 45, 45,   59,   59,   61,   61, 75, 75, 150, 150, 175, 175, 1000]
+#des   = [0, 0, 0,  1,  1,  1,  1,  0.1,  0.1,  0.1,  0.1,  1,  1,   1,   1,   0,   0,    0]
```
Comment (rgommers) [inline papers/scipy/code/signal/fir_filter_design_examples.py]: left over code to be removed

Comment (rgommers) [reply papers/scipy/code/signal/fir_filter_design_examples.py]: and more left over code further down in this file

File: papers/scipy/code/signal/fir_filter_design_examples.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+plt.figure(figsize=(8.0, 3.5))
+plt.plot(freqs, gains, 'k--', alpha=0.5, linewidth=1, label='ideal (firwin2)')
+
+for xlim, y in zip(np.reshape(rbands, (-1, 2)), rdes):
+    if xlim[0] == 0:
+        label="ideal (firls and remez)"
+    else:
+        label=None
+    #plt.plot(xlim, [y,y], 'k', alpha=0.25, linewidth=5, label=label)
```
Comment (rgommers) [inline papers/scipy/code/signal/fir_filter_design_examples.py]: with the plotting command commented out, this for-loop doesn't do anything

File: papers/scipy/code/signal/firlp_lowpass_example.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
+from __future__ import division, print_function
```
Comment (rgommers) [inline papers/scipy/code/signal/firlp_lowpass_example.py:1]: Maybe good to add this to all files consistently?

File: papers/scipy/code/signal/firlp_lowpass_example.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+
+        minimize
+            c.dot(x)
+        subject
+            A_ub.dot(x) <= b_ub
+            A_eq.dot(x) == b_eq
+
+    and convert the solution `x` to the FIR filter coefficients.
+
+    Warnings generated by `scipy.optimize.linprog` are converted to errors.
```
Comment (rgommers) [inline papers/scipy/code/signal/firlp_lowpass_example.py:22]: Maybe note here that this function requires SciPy 1.0

File: papers/scipy/code/signal/firlp_lowpass_example.py [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+        raise RuntimeWarning('linprog warning converted to error: %s' %
+                             (linprog_warning.args,))
+
+    if not result.success:
+        raise RuntimeError('linprog failed: %s' % (result.message,))
+
+    # Convert the solution to the linear programming problem to
+    # the FIR filter coefficients.
+    p = result.x[:-1]
+    taps = 0.5*np.concatenate((p[:0:-1], [2*p[0]], p[1:]))
```
Comment (rgommers) [inline papers/scipy/code/signal/firlp_lowpass_example.py]: This deserves a comment, hard to parse.

File: papers/scipy/code/signal/firlp_lowpass_example.py [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+    # The array of weights in the linear programming problem.
+    weights = np.concatenate((np.full_like(wpgrid, fill_value=1/deltap),
+                              np.full_like(wsgrid, fill_value=1/deltas)))
+    # The array of desired frequency responses.
+    desired = np.concatenate((np.ones_like(wpgrid),
+                              np.zeros_like(wsgrid)))
+
+    R = (numtaps - 1)//2
+    C = np.cos(wgrid[:, np.newaxis]*np.arange(R+1))
```
Comment (rgommers) [inline papers/scipy/code/signal/firlp_lowpass_example.py]: nitpick: spaces around ``*``

File: papers/scipy/code/signal/firwin2_examples.py [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+
+bands = [0, 48,   59,   61, 72, 150, 175, 1000]
+des   = [1,  1,  0.1,  0.1,  1,   1,   0,    0]
+taps_ls = firls(numtaps, bands, des, nyq=0.5*fs)
+
+rbands = [0, 48, 59, 61, 72, 150, 175, 1000]
+rdes   = [    1,    0.1,       1,         0]
+rwt    = [    3,      1,       3,         1]
+taps_r = remez(numtaps, rbands, rdes, Hz=fs)
+#taps_rw = remez(numtaps, rbands, rdes, weight=rwt, Hz=fs)
```
Comment (rgommers) [inline papers/scipy/code/signal/firwin2_examples.py]: left over code

File: papers/scipy/code/signal/kaiser_lowpass_filter_design.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+         linewidth=1, alpha=0.4)
+plt.plot([upper_trans, upper_trans], [-stop_db, upper_ripple], color='r',
+         linewidth=1, alpha=0.4)
+
+plt.ylim(-1.8*stop_db, 10)
+plt.ylabel('Gain (dB)')
+plt.title('Kaiser Window Filter Design', fontsize=10)
+
+#           'Inputs: width %g Hz, stop band rejection %.1f dB  '
+#           '(δ = %g)\n'
```
Comment (rgommers) [inline papers/scipy/code/signal/kaiser_lowpass_filter_design.py]: This file doesn't run for me:

    SyntaxError: Non-ASCII character '\xce' in file kaiser_lowpass_filter_design.py on line

File: papers/scipy/code/signal/opt_lowpass.py [hunk 13]
```diff
@@ -0,0 +1,153 @@
+
+import numpy as np
+from scipy.fftpack import next_fast_len
```
Comment (rgommers) [inline papers/scipy/code/signal/opt_lowpass.py]: unused import ``next_fast_len``

File: papers/scipy/code/signal/opt_lowpass.py [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
+from scipy.signal import remez, freqz
+import matplotlib.pyplot as plt
+
+
+def bellanger_estimate(deltap, deltas, width, fs):
+    """
+    Estimate the number of taps required for the given filter specifications.
+    """
+    n = (-2/3)*np.log10(10*deltap*deltas)*fs/width
```
Comment (rgommers) [inline papers/scipy/code/signal/opt_lowpass.py:14]: Integer division issue, this needs the ``from __future__`` import

Comment (rgommers) [reply papers/scipy/code/signal/opt_lowpass.py:14]: and a few spaces to appease the pep8 gods

File: papers/scipy/code/signal/opt_lowpass.py [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+    n = int(np.ceil(n))
+    return n
+
+
+def remez_lowpass(deltap, deltas, cutoff, width, fs):
+
+    numtaps = bellanger_estimate(deltap, deltas, width, fs)
+    numtaps |= 1  # Ensure an odd number of taps.
```
Comment (rgommers) [inline papers/scipy/code/signal/opt_lowpass.py]: I had to look up this operator:) Maybe say "bitwise OR" in a comment

File: papers/scipy/code/signal/opt_lowpass.py [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+
+# Frequency values in Hz
+fs = 1000.0
+cutoff = 180.0
+width = 30.0
+# Desired pass band ripple and stop band attenuation
+deltap = 0.005
+deltas = 0.002
+
+print("Pass band:  1 ± %g ([%.3g, %.3g] dB)" %
```
Comment (rgommers) [inline papers/scipy/code/signal/opt_lowpass.py]: Fails to run as well:

    SyntaxError: Non-ASCII character '\xc2' in file opt_lowpass.py on line 42, but no encoding declared

File: papers/scipy/code/signal/opt_lowpass.py [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+plt.plot(w, 20*np.log10(np.abs(h)))
+
+plt.plot([0, cutoff_lower_trans], [upper_ripple_db, upper_ripple_db], 'r', alpha=0.4)
+plt.plot([0, cutoff_lower_trans], [lower_ripple_db, lower_ripple_db], 'r', alpha=0.4)
+
+plt.plot([cutoff_upper_trans, 0.5*fs], [-stop_db, -stop_db], 'r', alpha=0.4)
+
+plt.axvline(cutoff_lower_trans, color='k', alpha=0.4, linewidth=1)
+plt.axvline(cutoff_upper_trans, color='k', alpha=0.4, linewidth=1)
+#plt.axvline(cutoff, color='c', alpha=0.25, linewidth=1)
```
Comment (rgommers) [inline papers/scipy/code/signal/opt_lowpass.py]: left over code

File: papers/scipy/code/signal/pressure_example.py [hunk 18]
```diff
@@ -0,0 +1,77 @@
+from __future__ import division, print_function
+
+import numpy as np
+import matplotlib.pyplot as plt
+from scipy.signal import sosfiltfilt, spectrogram
+from butter import butter_lowpass, butter_lowpass_filtfilt
```
Comment (rgommers) [inline papers/scipy/code/signal/pressure_example.py]: ``sosfiltfilt`` and ``butter_lowpass`` unused

File: papers/scipy/code/signal/remez_example.py [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+    w2, h2 = freqz(taps2, worN=8000)
+    w2 *= 0.5*fs/np.pi
+
+    plt.figure(figsize=(4.0, 3.0))
+    #for band, des in zip(np.array(bands).reshape(-1, 2), desired):
+    #    if band[0] == 0:
+    #        label = 'desired'
+    #    else:
+    #        label = None
+    #    plt.plot(band, [des]*2, 'g', alpha=0.2, linewidth=2.5, label=label)
```
Comment (rgommers) [inline papers/scipy/code/signal/remez_example.py]: left over code

**Review by rgommers — CHANGES_REQUESTED** [review]
Overall code LGTM. Have now reviewed all the scripts, will go through the paper content later.

**Comment by WarrenWeckesser** [comment]
Thanks, Ralf.  I pushed a new commit that cleans up the code.

File: papers/scipy/scipy.rst [hunk 20]
```diff
@@ -0,0 +1,1706 @@
...
+sparse matrix, interpolation, numerical integration, special functions
+
+.. contents::
+
+Introduction
+============
+
+[TO DO...]
+
+Related projects
```
Comment (rgommers) [inline papers/scipy/scipy.rst]: Do we need to include this, and if so does it need a separate section header? 

https://github.com/python-control/python-control is probably more relevant than ``statsmodels.tsa``

File: papers/scipy/scipy.rst [hunk 21]
```diff
@@ -0,0 +1,1706 @@
...
+----------------
+
+* ``statsmodels.tsa``: Time series analysis module in ``statsmodels``
+* others...
+
+There are two main classes of linear filters: *finite impulse response* (FIR)
+filters, and *infinite impulse response* (IIR) filters. 
+In the following two sections, we will discuss many of the functions
+that SciPy provides for the design and analysis of both types of filters.
```
Comment (rgommers) [inline papers/scipy/scipy.rst:58]: Probably also mention here spectral analysis (``spectrogram``). And perhaps even ``linprog``?

File: papers/scipy/scipy.rst [hunk 22]
```diff
@@ -0,0 +1,1706 @@
...
+We can compute the step response of this filter by applying it to
+an array of ones::
+
+    >>> x = np.ones(125)
+    >>> y = lfilter(b, a, x)
+    >>> plt.plot(y)
+
+The plot is shown in Figure :ref:`fig-unstable-butterworth`.
+(We haven't shown all the additional ``matplotlib`` function calls that
+we used to annotate the plot.)
```
Comment (rgommers) [inline papers/scipy/scipy.rst:251]: Maybe this comment can go together with the one about ``setprintoptions`` for IPython near the start of the chapter, and refer to the companion website or Github repo where all full examples can be found.

File: papers/scipy/scipy.rst [hunk 23]
```diff
@@ -0,0 +1,1706 @@
...
+            0.955,  0.964,  0.964,  0.988,  0.988])
+
+We can check the frequency response using ``scipy.signal.sosfreqz``::
+
+    >>> w, h = sosfreqz(sos, worN=8000)
+    >>> plt.plot(w/np.pi, np.abs(h))
+    [<matplotlib.lines.Line2D at 0x109ae9550>]
+    >>> plt.grid(alpha=0.25)
+    >>> plt.xlabel('Normalized frequency')
+    >>> plt.ylabel('Gain')
```
Comment (rgommers) [inline papers/scipy/scipy.rst:301]: Not sure if you did this on purpose, but this is half the plotting code for Fig 2 (there are 2 subplots in the paper). Perhaps confusing.

File: papers/scipy/scipy.rst [hunk 24]
```diff
@@ -0,0 +1,1706 @@
...
+   *Bottom*: Spectrogram of the filtered time series (generated using a
+   window size of 1.6 milliseconds).
+   The dashed line is at 1250 Hz.
+   :label:`fig-pressure-example-filtered`
+
+**Comments on creating a spectrogram.**
+The spectrograms in Figures :ref:`fig-pressure-example-input`
+and :ref:`fig-pressure-example-filtered` were created using ``spectrogram``
+from ``scipy.signal`` and ``pcolormesh`` from ``matplotlib.pyplot``.
+The function ``spectrogram`` has a several options that control how
```
Comment (rgommers) [inline papers/scipy/scipy.rst]: typo ``has a several``

File: papers/scipy/scipy.rst [hunk 25]
```diff
@@ -0,0 +1,1706 @@
...
+
+   *Top*: Filtered pressure, for the interval 15 < t < 35 (milliseconds).
+   The light gray curve is the unfiltered data.
+   *Bottom*: Spectrogram of the filtered time series (generated using a
+   window size of 1.6 milliseconds).
+   The dashed line is at 1250 Hz.
+   :label:`fig-pressure-example-filtered`
+
+**Comments on creating a spectrogram.**
+The spectrograms in Figures :ref:`fig-pressure-example-input`
```
Comment (rgommers) [inline papers/scipy/scipy.rst]: Perhaps worth spending one sentence on what a spectrogram is? Also can mention ``signal.stft``, creates the same type of spectrogram.

File: papers/scipy/scipy.rst [hunk 26]
```diff
@@ -0,0 +1,1706 @@
...
+
+In this example, we will use synthetic data to demonstrate a
+bandpass filter.  We have 0.03 seconds of data sampled at
+4800 Hz.  We want to apply a bandpass filter to remove frequencies
+below 400 Hz or above 1200 Hz.
+
+Just like we did for the lowpass filter, we define two functions that
+allow us to create and apply a Butterworth bandpass filter with the
+frequencies given in Hz (or any other units).  The functions take care
+of scaling the values to the normalized range expected by the SciPy function
```
Comment (rgommers) [inline papers/scipy/scipy.rst]: ``the SciPy function`` seems a bit redundant

File: papers/scipy/scipy.rst [hunk 27]
```diff
@@ -0,0 +1,1706 @@
...
+    :label:`fig-bandpass-example-signals`
+
+Filtering a long signal in batches
+----------------------------------
+
+We will again use synthetic data generated by the same function
+used in the previous example, but for a longer time interval.
+
+This example shows how the state of the IIR filter can be saved
+and restored, so a filter can be applied to a long signal in batches.
```
Comment (rgommers) [inline papers/scipy/scipy.rst]: It wasn't immediately clear to me why this was done - first thought was to speed up the calculation, but it's because you want to filter a signal while the data is still being acquired (right?). Worth pointing out.

File: papers/scipy/scipy.rst [hunk 28]
```diff
@@ -0,0 +1,1706 @@
...
+for ``y`` and ``x``.
+
+Putting it all together, here is the code using ``lfilter`` that
+replaces the for-loop shown above::
+
+    b = [0, 1]
+    a = [1, -beta]
+    zi = lfiltic(b, a, y=[0], x=[alpha])
+    y, zo = lfilter(b, a, alpha + gamma*h, zi=zi)
+
```
Comment (rgommers) [inline papers/scipy/scipy.rst:750]: Suggest to conclude with ``Result: X times faster for a signal of size N.``

File: papers/scipy/scipy.rst [hunk 29]
```diff
@@ -0,0 +1,1706 @@
...
+applies the Savitzky-Golay filter to an input signal without returning the
+filter coefficients.
+
+FIR filter frequency response
+-----------------------------
+
+The function ``scipy.signal.freqz`` computes the frequency
+response of a FIR filter (and, more generally, an IIR filter represented
+by the coefficients of the numerator and denominator of the transfer
+function).
```
Comment (rgommers) [inline papers/scipy/scipy.rst]: Not sure I get the point of the part in brackets. Raises user questions like "why is IIR more general" and "why does it matter that that filter is in num/den format?

File: papers/scipy/scipy.rst [hunk 30]
```diff
@@ -0,0 +1,1706 @@
...
+* *The window method.*
+
+  Two functions implement the window method (see, for example,
+  Oppenheim and Schafer [OS]_, Ch. 7.5):
+
+  * ``scipy.signal.firwin``
+  * ``scipy.signal.firwin2``
+
+  We'll show an example of ``firwin2`` in the following section,
+  and we'll use ``firwin`` in when we discuss the Kaiser window method.
```
Comment (rgommers) [inline papers/scipy/scipy.rst]: typo ``in when``

File: papers/scipy/scipy.rst [hunk 31]
```diff
@@ -0,0 +1,1706 @@
...
+   Note the possibly undesirable behavior of the filter in the
+   transition interval [550, 700].
+
+   :label:`fig-remez-example-47taps`
+
+In some cases, the exchange algorithm implemented in ``remez`` can fail
+to converge.  This can happen, for example, when the transition between
+bands or the bands themselves are too small.  Failure is more likely when
+the number of taps is large (i.e. greater than 1000).  When the algorithm
+fails to converge, ``remez`` raises an exception, so the failure can be
```
Comment (rgommers) [inline papers/scipy/scipy.rst]: This sentence could be left out.

Comment (rgommers) [reply papers/scipy/scipy.rst]: Looking at the rest of the section, the discussion on failures can be shortened significantly I think. Perhaps just say that for >1000 taps there may be issues and recommend to check the frequence response.

**Review by rgommers — CHANGES_REQUESTED** [review]
Overall this reads well! My comments are all local, the structure looks good to me.

**Comment by WarrenWeckesser** [comment]
Ralf, I just pushed a commit that updates the sections "FIR filter design" and "FIR filter design: the window method".  I'll get back to your recent comments in a little while.

File: papers/scipy/scipy.rst (hunk 20 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst]: Removed.

File: papers/scipy/scipy.rst (hunk 21 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst:58]: Sounds good, but I'm going to defer work on editing the introduction until after the rest of the paper has settled a bit more.

File: papers/scipy/scipy.rst (hunk 22 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst:251]: Yes, I was planning on something like that.  This was written before the later sections.  You can see that later, I switch to just giving the essential code instead of showing commands in an interactive session.  On my to-do list is to update this section to use a similar style.  That change might make it unnecessary to fiddle with `set_printoptions`.

File: papers/scipy/scipy.rst (hunk 23 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst:301]: As I mentioned in another comment, I think these transcripts from interactive sessions will be changed to show just the essential functon calls.

File: papers/scipy/scipy.rst (hunk 25 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst]: I added a sentence about what a spectrogram is.  I'm not sure about also mentioning `stft`, so I didn't.

File: papers/scipy/scipy.rst (hunk 24 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst]: Fixed.

File: papers/scipy/scipy.rst (hunk 27 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst]: I edited this section to (hopefully) clarify the motivation.

File: papers/scipy/scipy.rst (hunk 28 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst:750]: Not really needed--the motivation isn't speed.

File: papers/scipy/scipy.rst (hunk 29 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst]: Edited to be clearer (I hope!).

File: papers/scipy/scipy.rst (hunk 31 above)
Comment (WarrenWeckesser) [reply papers/scipy/scipy.rst]: OK, shortened.

**Comment by WarrenWeckesser** [comment]
Ralf, I addressed your latest comments.

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
Thanks Warren, changes LGTM. Merging.
